### PR TITLE
CMake for crossplatform support

### LINKED ISSUE #32: [Feature] Investigate compilation using CMake for linux systems

### PR DESCRIPTION
**Pull request by wolffshots**
closes #32 

## Some examples of CMake usage

The build output is condensed since it is already generated and compiled before showing these

### Linux (x86) example

image

### macOS (arm) example

image

### Windows (x86) example

image

### WSL (x86) example

image

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,17 +1,50 @@
...
-		${SOURCE_DIR}/WatchdogChunk.cpp
-	INCLUDE_DIRS
-		${INCLUDE_DIR}
-)
\ No newline at end of file
+# Set C++ standard
+set(CMAKE_CXX_STANDARD 11)
+
+# Main executable
+add_library(ChunkTypesLib
```
Comment (Grabt234) [inline CMakeLists.txt:8]: You can define variables in cmake. As an example:

set(MY_VARIABLE "Hello, CMake!")

and then call it using:

${MY_VARIABLE}

I would suggest using varibales like SOURCE_DIR, SOURCE_DIR_TESTS
This should allow for quick changes to cmake if the repository structure ever changes

Comment (Grabt234) [reply CMakeLists.txt:8]: An example of how this can be done is in the esp folder

File: .github/workflows/cmake.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# This workflow uses actions that are not certified by GitHub.
```
Comment (Grabt234) [inline .github/workflows/cmake.yml:1]: This is just for my sanity:

Name: The name of the workflow

On: When the workflow occurs

Permissions: What the workflow is allowed to do

Jobs: What the workflow shall do

Build: What env is used

Steps: What is done to complete the job

Generate: Create the make files (???)

Build: Use the make files to compile the code

Test: Run the tests

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,17 +1,50 @@
+cmake_minimum_required(VERSION 3.10)
+project(ChunkTypesProject CXX)
 
-# Min CMake version
```
Comment (Grabt234) [inline CMakeLists.txt:2]: Not entirely sure how the ESP dir gets used now

File: .github/workflows/cmake.yml (hunk 2 above)
Comment (wolffshots) [reply .github/workflows/cmake.yml:1]: aye aye cap'n

File: CMakeLists.txt (hunk 3 above)
Comment (wolffshots) [reply CMakeLists.txt:2]: it doesn’t right now, need to set my espidf back up and make a test project to see how i can import and use Chunk_Types

File: CMakeLists.txt (hunk 1 above)
Comment (wolffshots) [reply CMakeLists.txt:13]: cool, have implemented some vars for dir names

File: CMakeLists.txt (hunk 3 above)
Comment (wolffshots) [reply CMakeLists.txt:2]: now it does like so:
```cmake
# generic setup

set(SOURCE_DIR "source")
set(INCLUDE_DIR "include")
set(SOURCE_TESTS_DIR "source_tests")
set(INCLUDE_TESTS_DIR "include_tests”)

if(DEFINED ENV{ESP_PLATFORM})
    message(STATUS "Building for ESP32")

    idf_component_register(
        SRCS
        ...
    )
    
    # esp specific config

else()
    message(STATUS "Building for Unix")
    
    # unix specific setup

endif()
```

Comment (wolffshots) [reply CMakeLists.txt:2]: still not a perfect automated system since it requires you to set a var before building for esp

Comment (wolffshots) [reply CMakeLists.txt:2]: okay i’ve updated it to check for `IDF_PATH` rather which should only be present in the ESP IDF terminal or after specifically running the export function

File: .gitignore [hunk 4]
```diff
@@ -9,4 +9,6 @@ x64/
 
 # Other
 build/
-build_ESP32/
\ No newline at end of file
+build_ESP32/
+
+.DS_Store
```
Comment (Grabt234) [inline .gitignore:14]: This is a config folder for MAC

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,17 +1,75 @@
...
 cmake_minimum_required(VERSION 3.20)
+project(ChunkTypesProject CXX)
 
+# Set directory vars
 set(SOURCE_DIR "source")
 set(INCLUDE_DIR "include")
+set(SOURCE_TESTS_DIR "source_tests")
+set(INCLUDE_TESTS_DIR "include_tests")
+
+if(DEFINED ENV{IDF_PATH}) # this is normally only set when running in an esp idf terminal
```
Comment (Grabt234) [inline CMakeLists.txt:10]: If IDF PATH is defined, then the build system will use esp idf

**Review by Grabt234 — APPROVED** [review]

File: .gitignore (hunk 4 above)
Comment (wolffshots) [reply .gitignore:14]: *file but yes

**Pull request merged**